Release 1.8.0

MichaelStritt opened the pull request:
### Linked issue

Release PR, related issue #805.

### How to test

Check out testing descriptions in #805.

### Comments

Starting to test now, discussion (& release) on monday (if possible).

MichaelStritt commented:
There were some issues during unit testing. Will look into it...

```java
====================================== TEST RESULTS ======================================

                        name                           unit           tests        passed
    ____________________________________________    __________    _____________    ______

    'xASL_adm_CatchNumbersFromString'               'function'    [1×3  struct]    true  
    'xASL_adm_CheckFileCount'                       'function'    [1×2  struct]    true  
    'xASL_adm_CompareLists'                         'function'    [1×2  struct]    true  
    'xASL_adm_CorrectName'                          'function'    [1×3  struct]    true  
    'xASL_adm_DeleteFileList'                       'function'    [1×2  struct]    true  
    'xASL_adm_OrderFields'                          'function'    [1×1  struct]    true  
    'xASL_bids_BIDS2Legacy'                         'function'    [1×2  struct]    false 
    'xASL_bids_CheckDatasetDescription'             'function'    [1×2  struct]    true  
    'xASL_bids_Config'                              'function'    [1×1  struct]    true  
    'xASL_bids_CreateDatasetDescriptionTemplate'    'function'    [1×2  struct]    true  
    'xASL_bids_JsonCheck'                           'function'    [1×3  struct]    true  
    'xASL_bids_VendorFieldCheck'                    'function'    [1×1  struct]    true  
    'xASL_io_ExportVTK'                             'function'    [1×5  struct]    true  
    'xASL_io_Nifti2Im'                              'function'    [1×1  struct]    true  
    'xASL_stat_PSNR'                                'function'    [1×2  struct]    true  
    'xASL_stat_QuantileNan'                         'function'    [1×1  struct]    true  
    'xASL_stat_StdNan'                              'function'    [1×3  struct]    true  
    'xASL_str2num'                                  'function'    [1×3  struct]    true  
    'xASL_test_GetLogContent'                       'function'    [1×2  struct]    true  
    'xASL_tsvRead'                                  'function'    [1×2  struct]    true  
    'xASL_tsvWrite'                                 'function'    [1×2  struct]    true  
    'xASL_ExploreASL_Master'                        'master'      [1×11 struct]    false 
    'xASL_module_ASL'                               'module'      [1×1  struct]    true  
    'xASL_module_Population'                        'module'      [1×1  struct]    true  
    'xASL_module_Structural'                        'module'      [1×1  struct]    true  

==========================================================================================
```

MichaelStritt commented:
Second time doing the unit testing, after commit a8575e9 everything seems to work just fine ...

```java
====================================== TEST RESULTS ======================================

                        name                           unit           tests        passed
    ____________________________________________    __________    _____________    ______

    'xASL_adm_CatchNumbersFromString'               'function'    [1×3  struct]    true  
    'xASL_adm_CheckFileCount'                       'function'    [1×2  struct]    true  
    'xASL_adm_CompareLists'                         'function'    [1×2  struct]    true  
    'xASL_adm_CorrectName'                          'function'    [1×3  struct]    true  
    'xASL_adm_DeleteFileList'                       'function'    [1×2  struct]    true  
    'xASL_adm_OrderFields'                          'function'    [1×1  struct]    true  
    'xASL_bids_BIDS2Legacy'                         'function'    [1×2  struct]    true  
    'xASL_bids_CheckDatasetDescription'             'function'    [1×2  struct]    true  
    'xASL_bids_Config'                              'function'    [1×1  struct]    true  
    'xASL_bids_CreateDatasetDescriptionTemplate'    'function'    [1×2  struct]    true  
    'xASL_bids_JsonCheck'                           'function'    [1×3  struct]    true  
    'xASL_bids_VendorFieldCheck'                    'function'    [1×1  struct]    true  
    'xASL_io_ExportVTK'                             'function'    [1×5  struct]    true  
    'xASL_io_Nifti2Im'                              'function'    [1×1  struct]    true  
    'xASL_stat_PSNR'                                'function'    [1×2  struct]    true  
    'xASL_stat_QuantileNan'                         'function'    [1×1  struct]    true  
    'xASL_stat_StdNan'                              'function'    [1×3  struct]    true  
    'xASL_str2num'                                  'function'    [1×3  struct]    true  
    'xASL_test_GetLogContent'                       'function'    [1×2  struct]    true  
    'xASL_tsvRead'                                  'function'    [1×2  struct]    true  
    'xASL_tsvWrite'                                 'function'    [1×2  struct]    true  
    'xASL_ExploreASL_Master'                        'master'      [1×11 struct]    true  
    'xASL_module_ASL'                               'module'      [1×1  struct]    true  
    'xASL_module_Population'                        'module'      [1×1  struct]    true  
    'xASL_module_Structural'                        'module'      [1×1  struct]    true  

==========================================================================================
```

MichaelStritt commented:
> 
> 
> @MichaelStritt This is incorrect, and tries to fix symptoms rather than the root ;)

I wasn't fixing symptoms. We just didn't do it for single sessions like that before.
I updates the unit test and fixed the isempty statement. I think you got it a bit wrong.
Please check commit 69680e2 👍

MichaelStritt commented:
Okay, so for the `ExploreASL_Master` unit test, there is only one subtest (number 8) that went wrong. Probably I just forgot to update one of those paths for the test condition check. Interesting that the second subtest of `xASL_stat_StdNan` failed. Probably the variability is a bit higher than I thought it would be.

```java
====================================== TEST RESULTS ======================================

                        name                           unit           tests        passed
    ____________________________________________    __________    _____________    ______

    'xASL_adm_CatchNumbersFromString'               'function'    [1×3  struct]    true  
    'xASL_adm_CheckFileCount'                       'function'    [1×2  struct]    true  
    'xASL_adm_CompareLists'                         'function'    [1×2  struct]    true  
    'xASL_adm_CorrectName'                          'function'    [1×3  struct]    true  
    'xASL_adm_DeleteFileList'                       'function'    [1×2  struct]    true  
    'xASL_adm_OrderFields'                          'function'    [1×1  struct]    true  
    'xASL_bids_BIDS2Legacy'                         'function'    [1×2  struct]    true  
    'xASL_bids_CheckDatasetDescription'             'function'    [1×2  struct]    true  
    'xASL_bids_Config'                              'function'    [1×1  struct]    true  
    'xASL_bids_CreateDatasetDescriptionTemplate'    'function'    [1×2  struct]    true  
    'xASL_bids_JsonCheck'                           'function'    [1×3  struct]    true  
    'xASL_bids_VendorFieldCheck'                    'function'    [1×1  struct]    true  
    'xASL_io_ExportVTK'                             'function'    [1×5  struct]    true  
    'xASL_io_Nifti2Im'                              'function'    [1×1  struct]    true  
    'xASL_stat_PSNR'                                'function'    [1×2  struct]    true  
    'xASL_stat_QuantileNan'                         'function'    [1×1  struct]    true  
    'xASL_stat_StdNan'                              'function'    [1×3  struct]    false 
    'xASL_str2num'                                  'function'    [1×3  struct]    true  
    'xASL_test_GetLogContent'                       'function'    [1×2  struct]    true  
    'xASL_tsvRead'                                  'function'    [1×2  struct]    true  
    'xASL_tsvWrite'                                 'function'    [1×2  struct]    true  
    'xASL_ExploreASL_Master'                        'master'      [1×11 struct]    false 
    'xASL_module_ASL'                               'module'      [1×1  struct]    true  
    'xASL_module_Population'                        'module'      [1×1  struct]    true  
    'xASL_module_Structural'                        'module'      [1×1  struct]    true  

==========================================================================================
```

MichaelStritt commented:
@jan-petr: super minor bug in the unit testing, should be fixed 221dd78 👍

MichaelStritt commented:
### Flavor Tests

BIDS conversion comparison:

```
Dataset: GE_PCASL_2DEPI_23.0LX_1
Dataset: GE_PCASL_3Dspiral_14.0LX_1
Dataset: GE_PCASL_3Dspiral_15.0LX_WIP_1
Dataset: GE_PCASL_3Dspiral_22.0LX_1
Dataset: GE_PCASL_3Dspiral_23.1LX_1
Dataset: GE_PCASL_3Dspiral_24.0LX_WIP_1
Dataset: GE_PCASL_3Dspiral_25.0LX_ADNI3_027_S_5079_1
Dataset: Philips_PASL_2DEPI_5.4.1.0_ADNI3_006_S_6681_1
Dataset: Philips_PCASL_2DEPI_3.2.1.1_1
Dataset: Philips_PCASL_2DEPI_3.2.1.1_2
Dataset: Philips_PCASL_2DEPI_3.2.1.1_3
Dataset: Philips_PCASL_2DEPI_3.2.1.1_dummyLL_1
Dataset: Philips_PCASL_2DEPI_3.2.1.1_dummyMultiPLD_1
Dataset: Philips_PCASL_2DEPI_3.2.1.1_dummyQUASAR_1
Dataset: Philips_PCASL_2DEPI_3.2.2.0_glioma_1
Dataset: Philips_PCASL_2DEPI_3.2.2.1_1
Dataset: Philips_PCASL_2DEPI_3.2.3.0_BSupN_1
Dataset: Philips_PCASL_2DEPI_3.2.3.1_1
Dataset: Philips_PCASL_2DEPI_3.2.3.2_1
Dataset: Philips_PCASL_2DEPI_4.1.2.1_1
Dataset: Philips_PCASL_2DEPI_5.1.7.2_1
Dataset: Philips_PCASL_2DEPI_5.1.7.2_2
Dataset: Philips_PCASL_2DEPI_5.1.8.2_1
Dataset: Philips_PCASL_2DEPI_5.1.9.1_GMB_1
Dataset: Philips_PCASL_2DEPI_5.3.1.1_1
Dataset: Philips_PCASL_2DEPI_PARREC_1
Dataset: Philips_PCASL_3DGRASE_5.1.7.2_1
Dataset: Philips_PCASL_3DGRASE_5.4.0.2_1
Dataset: Philips_PCASL_3DGRASE_5.4.1.0_TopUp_1
...
```

No warnings 😃 

```
...
Dataset: Siemens_PASL_2DEPI_VB15A_ADNI2_941_S_4764_1
Dataset: Siemens_PASL_2DEPI_VB17A_ADNI2_941_S_5193_1
Dataset: Siemens_PASL_2DEPI_VB17A_BSupN_1
Dataset: Siemens_PASL_2DEPI_VB17A_BSupN_2
Dataset: Siemens_PASL_2DEPI_VB19A_BSupN_1
Dataset: Siemens_PASL_3DGRASE_E11_1
Dataset: Siemens_PASL_3DGRASE_VB17A_1
Dataset: Siemens_PASL_3DGRASE_VD13D_1
Dataset: Siemens_PASL_3DGRASE_VE11C_1
Dataset: Siemens_PASL_3DGRASE_VE11C_2
Dataset: Siemens_PCASL_2DEPI_NA_1
Dataset: Siemens_PCASL_2DEPI_VB17A_BSupN_1
File:      \sub-Sub1\perf\sub-Sub1_asl.nii.gz
           RMSE of NIFTIs above threshold.
Warning: RMSE of NIFTIs above threshold: \sub-Sub1\perf\sub-Sub1_asl.nii.gz    
> In xASL_bids_CompareStructures>printDifferencesAsWarnings (line 217)
  In xASL_bids_CompareStructures (line 206)
  In xASL_test_Flavors (line 137)
  In xASL_test_FullPipelineTest (line 66) 

Dataset: Siemens_PCASL_2DEPI_VB17A_BSupN_2
Dataset: Siemens_PCASL_3DGRASE_VB17A_TopUp_1
Dataset: Siemens_PCASL_3DGRASE_VB17A_multiPLD_1
Dataset: Siemens_PCASL_3DGRASE_VD13A_Hadamard4-FME_1
Dataset:   rawdata_A
Missing:   \sub-Sub1\perf\sub-Sub1_m0scan.json
Missing:   \sub-Sub1\perf\sub-Sub1_m0scan.nii.gz
File:      \sub-Sub1\perf\sub-Sub1_asl.json
           Extra field: BackgroundSuppressionNumberPulses
           Extra field: BackgroundSuppressionPulseTime
           Extra field: EchoNumber
           Extra field: LabelingDuration
           Different dimension: EchoTime
           Different value: M0Type (Absent vs Separate)
File:      \sub-Sub1\perf\sub-Sub1_asl.nii.gz
           RMSE of NIFTIs above threshold.
File:      dataset_description.json
           Different value: Name (Siemens_PCASL_3DGRASE_VD13A_Hadamard4-FME_1 vs Siemens_Hadamard4_3DGRASE_FME)
Warning: Missing file \sub-Sub1\perf\sub-Sub1_m0scan.json   
> In xASL_bids_CompareStructures>printMissingAsWarnings (line 240)
  In xASL_bids_CompareStructures (line 205)
  In xASL_test_Flavors (line 137)
  In xASL_test_FullPipelineTest (line 66) 

Warning: Missing file \sub-Sub1\perf\sub-Sub1_m0scan.nii.gz   
> In xASL_bids_CompareStructures>printMissingAsWarnings (line 240)
  In xASL_bids_CompareStructures (line 205)
  In xASL_test_Flavors (line 137)
  In xASL_test_FullPipelineTest (line 66) 

Warning: Different file content: \sub-Sub1\perf\sub-Sub1_asl.json    
> In xASL_bids_CompareStructures>printDifferencesAsWarnings (line 217)
  In xASL_bids_CompareStructures (line 206)
  In xASL_test_Flavors (line 137)
  In xASL_test_FullPipelineTest (line 66) 

Warning: RMSE of NIFTIs above threshold: \sub-Sub1\perf\sub-Sub1_asl.nii.gz    
> In xASL_bids_CompareStructures>printDifferencesAsWarnings (line 217)
  In xASL_bids_CompareStructures (line 206)
  In xASL_test_Flavors (line 137)
  In xASL_test_FullPipelineTest (line 66) 

Warning: Different file content: dataset_description.json    
> In xASL_bids_CompareStructures>printDifferencesAsWarnings (line 217)
  In xASL_bids_CompareStructures (line 206)
  In xASL_test_Flavors (line 137)
  In xASL_test_FullPipelineTest (line 66) 

Dataset: Siemens_PCASL_3DGRASE_VD13A_Hadamard8-FME_1
Dataset:   rawdata_A
Missing:   \sub-Sub1\perf\sub-Sub1_m0scan.json
Missing:   \sub-Sub1\perf\sub-Sub1_m0scan.nii.gz
File:      \sub-Sub1\perf\sub-Sub1_asl.json
           Extra field: BackgroundSuppressionNumberPulses
           Extra field: BackgroundSuppressionPulseTime
           Extra field: LabelingDuration
           Different dimension: EchoTime
           Different value: M0Type (Absent vs Separate)
Warning: Missing file \sub-Sub1\perf\sub-Sub1_m0scan.json   
> In xASL_bids_CompareStructures>printMissingAsWarnings (line 240)
  In xASL_bids_CompareStructures (line 205)
  In xASL_test_Flavors (line 137)
  In xASL_test_FullPipelineTest (line 66) 

Warning: Missing file \sub-Sub1\perf\sub-Sub1_m0scan.nii.gz   
> In xASL_bids_CompareStructures>printMissingAsWarnings (line 240)
  In xASL_bids_CompareStructures (line 205)
  In xASL_test_Flavors (line 137)
  In xASL_test_FullPipelineTest (line 66) 

Warning: Different file content: \sub-Sub1\perf\sub-Sub1_asl.json    
> In xASL_bids_CompareStructures>printDifferencesAsWarnings (line 217)
  In xASL_bids_CompareStructures (line 206)
  In xASL_test_Flavors (line 137)
  In xASL_test_FullPipelineTest (line 66) 

Dataset: Siemens_PCASL_3DGRASE_VD13D_1
Dataset: Siemens_PCASL_3DGRASE_VD13D_2
Dataset: Siemens_PCASL_3DGRASE_VE11C_1
Dataset: Siemens_PCASL_3DGRASE_VE11C_2
```

MichaelStritt commented:
- Hmm... not sure why I get this warning `Warning: Please use ASL4D for your ASL tokenScanAlias instead.` for `Siemens_PASL_2DEPI_VB17A_BSupN_2` & `Siemens_PCASL_3DGRASE_VB17A_multiPLD_1`, because the tokens actually seem fine
- For `Siemens_PCASL_3DGRASE_VD13A_Hadamard4-FME_1` we get `ExploreASL Import module failed... Too many input arguments. xASL_imp_DCM2NII_CheckIfTimeEncoded, line 172...`. This is probably okay for now, but would be nicer if we would instead terminate the pipeline in a controlled way here.
- `dataset_description.json`, different value: Name (`Siemens_PCASL_3DGRASE_VD13A_Hadamard4-FME_1 vs Siemens_Hadamard4_3DGRASE_FME`), that should be an easy one to fix as well

MichaelStritt commented:
## 10 test datasets

![image](https://user-images.githubusercontent.com/46593074/132237797-bb39ee35-769c-4cc6-ae74-3ad2e5e5954e.png)

![image](https://user-images.githubusercontent.com/46593074/132237956-abee3856-cfac-40b2-9b57-b15e28a0670f.png)

MichaelStritt commented:
# Flavor testing

## BIDS comparison

Here everything seems perfectly fine. The only warnings that pop up are related to the FME datasets and I guess we can ignore them for now or we just fix the reference.

```
>> xASL_test_Flavors(pathExploreASL, pathTest, [0 0 1 0 0 0 0], x);
...

Dataset: Siemens_PCASL_3DGRASE_VD13A_Hadamard4-FME_1
File:      \sub-Sub1\perf\sub-Sub1_asl.json
           Extra field: BackgroundSuppressionNumberPulses
           Extra field: BackgroundSuppressionPulseTime
           Extra field: EchoNumber
           Extra field: LabelingDuration
Warning: Different file content: \sub-Sub1\perf\sub-Sub1_asl.json    
> In xASL_bids_CompareStructures>printDifferencesAsWarnings (line 217)
  In xASL_bids_CompareStructures (line 206)
  In xASL_test_Flavors (line 137) 

Dataset: Siemens_PCASL_3DGRASE_VD13A_Hadamard8-FME_1
File:      \sub-Sub1\perf\sub-Sub1_asl.json
           Extra field: BackgroundSuppressionNumberPulses
           Extra field: BackgroundSuppressionPulseTime
           Extra field: LabelingDuration
Warning: Different file content: \sub-Sub1\perf\sub-Sub1_asl.json    
> In xASL_bids_CompareStructures>printDifferencesAsWarnings (line 217)
  In xASL_bids_CompareStructures (line 206)
  In xASL_test_Flavors (line 137) 
```

## Processing

We should definitely address these bugs...

### Philips_PASL_2DEPI_5.4.1.0_ADNI3_006_S_6681_1

`ASL module terminated for subject 1: sub-001 -> Reference to non-existent field 'LabelingDuration'.`
(xASL_quant_SinglePLD, line 139)

### Philips_PCASL_3DGRASE_5.1.7.2_1

`ASL module terminated for subject 1: sub-Patient1_1 -> Error`
(xASL_module_ASL, line 220)

### Siemens_PASL_3DGRASE_E11_1

`ASL module terminated for subject 1: sub-Sub1_1 -> Error`
(spm_realign>error_message, line 525)

### Siemens_PASL_3DGRASE_VE11C_2

`ASL module terminated for subject 1: sub-Sub1_1 -> Array indices must be positive integers or logical values`
(xASL_im_M0ErodeSmoothExtrapolate, line 77)

### Siemens_PCASL_3DGRASE_VB17A_TopUp_1

`ASL module terminated for subject 1: sub-Sub1_1 -> Error`
(xASL_spm_admin, line 29, xASL_spm_coreg, line 108)

### Siemens_PCASL_3DGRASE_VB17A_TopUp_2

`ASL module terminated for subject 1: sub-SUB002_1 -> Reference to non-existent field 'LabelingDuration'.`
(xASL_quant_SinglePLD, line 142, xASL_wrp_Quantify, line 365)

### Siemens_PCASL_3DGRASE_VB17A_multiPLD_1

`ASL module terminated for subject 1: sub-Sub1_1 -> Error`
(xASL_quant_BSupCalculation, line 102)

### Siemens_PCASL_3DGRASE_VD13A_Hadamard4-FME_1

`ASL module terminated for subject 1: sub-Sub1_1 -> Matrix dimensions must agree.`

### Siemens_PCASL_3DGRASE_VD13A_Hadamard8-FME_1

`ASL module terminated for subject 1: sub-Sub1_1 -> Operands to the || and && operators must be convertible to logical scalar values.`


## Summary

@jan-petr & @HenkMutsaerts 

I guess we can ignore the bugs in the Hadamard FME datasets, even though I thought we would terminate the processing in a more controlled way there instead of just letting ExploreASL crash while this is work in progress. If we can detect FME and we know the program can't handle it right now, we should stop the processing, right?

The same goes for the top up datasets. I got the feeling that we had issues with the top up processing since release 1.0.0 and if we know where it goes wrong, we should terminate ExploreASL in a more controlled way, right?

We have two datasets with missing labeling duration. Can we just fix this by updating the datasets studyPar in the flavor database?

I was kind of surprised by the errors that occurred within spm code. Do we have empty images or something like that again? Or do you know where this is coming from?

The one in the M0 Erode Smooth Extrapolate script is definitely problematic. This seems like a real programming bug. We probably need to check the image which is being modified to make sure that these conditions don't crash.

jan-petr commented:
Some errors noted above in the processing are indeed due to having empty images or missing crucial parameters - see #342 and update if needed. You have fixed some warnings already, but we need to improve the Flavors themselves as well. But this is really mostly flavors problem and can be solved later...

jan-petr commented:
`  Linux 1.6.0

    {'Data'             }    {'mean_qCBF_TotalGM'}    {'median_qCBF_Tot…'}    {'median_qCBF_Dee…'}    {'CoV_qCBF_TotalGM'}    {'GMvol' }    {'WMvol' }        {'CSFvol'}    {'PipelineCompleted'}    {'TC_ASL_Registra…'}    {'TC_M0_Registrat…'}
    {'GE_3Dspiral_T1p…'}    {'65.0911'          }    {'52.4651'          }    {'32.0559'          }    {'0.29921'         }    {'0.6400'}    {'0.5054'}   {'0.4330'}    {'1'                }    {'0.8995'           }    {'0.9315'           }
    {'Philips_2DEPI_F…'}    {'38.9729'          }    {'27.6587'          }    {'6.73000'          }    {'0.71513'         }    {'0.5349'}    {'0.4590'}   {'0.4056'}    {'1'                }    {'0.7292'           }    {'0.8137'           }
    {'Philips_2DEPI_F…'}    {'41.7854'          }    {'31.8535'          }    {'13.6237'          }    {'0.37658'         }    {'0.5442'}    {'0.3920'}      {'0.3561'}    {'1'                }    {'0.9021'           }    {'0.9479'           }
    {'Philips_2DEPI_F…'}    {'47.3228'          }    {'42.3709'          }    {'20.0149'          }    {'0.42938'         }    {'0.6994'}    {'0.5816'}      {'0.2222'}    {'1'                }    {'0.8157'           }    {'0.8454'           }
    {'Philips_2DEPI_L…'}    {'empty'            }    {'14.1728'          }    {'2.99130'          }    {'0.89536'         }    {'0.7096'}    {'0.5136'}      {'0.1942'}    {'1'                }    {'0.7593'           }    {'0.8911'           }
    {'Philips_2DEPI_S…'}    {'45.9512'          }    {'41.1554'          }    {'9.40980'          }    {'0.36523'         }    {'0.8167'}    {'0.4596'}      {'0.1516'}    {'1'                }    {'0.8544'           }    {'0.8028'           }
    {'Philips_3DGRASE…'}    {'35.1324'          }    {'26.2474'          }    {'7.62250'          }    {'0.66049'         }    {'0.5716'}    {'0.5282'}      {'0.4207'}    {'1'                }    {'0.8568'           }    {'0.7545'           }
    {'Philips_3DGRASE…'}    {'31.9461'          }    {'24.3224'          }    {'8.58920'          }    {'0.43733'         }    {'0.7239'}    {'0.5520'}      {'0.3564'}    {'1'                }    {'0.8389'           }    {'0.8932'           }
    {'Siemens3DGRASE_…'}    {'37.6957'          }    {'37.2872'          }    {'21.7901'          }    {'0.78873'         }    {'0.5191'}    {'0.4658'}      {'0.4054'}    {'1'                }    {'0.6915'           }    {'0.4987'           }
    {'Siemens_2DEPI_n…'}    {'67.5016'          }    {'50.6498'          }    {'13.4243'          }    {'0.31698'         }    {'0.6886'}    {'0.6030'}      {'0.3101'}    {'1'                }    {'0.8757'           }    {'0.9589'           }`

`1.7.0

    {'Data'             }    {'mean_qCBF_TotalGM'}    {'median_qCBF_Tot…'}    {'median_qCBF_Dee…'}    {'CoV_qCBF_TotalGM'}    {'GMvol' }    {'WMvol' }      {'CSFvol'}    {'PipelineCompleted'}    {'TC_ASL_Registra…'}    {'TC_M0_Registrat…'}
    {'GE_3Dspiral_T1p…'}    {'65.092'           }    {'52.4463'          }    {'31.8969'          }    {'0.29939'         }    {'0.6399'}    {'0.5052'}      {'0.4329'}    {'1'                }    {'0.9005'           }    {'0.9312'           }
    {'Philips_2DEPI_F…'}    {'39.1474'          }    {'27.7875'          }    {'6.7287'           }    {'0.71379'         }    {'0.5349'}    {'0.4590'}      {'0.4056'}    {'1'                }    {'0.7296'           }    {'0.8141'           }
    {'Philips_2DEPI_F…'}    {'41.7268'          }    {'31.8413'          }    {'13.6215'          }    {'0.37679'         }    {'0.5442'}    {'0.3921'}    {'0.3564'}    {'1'                }    {'0.9021'           }    {'0.9480'           }
    {'Philips_2DEPI_F…'}    {'49.2752'          }    {'43.7215'          }    {'19.7391'          }    {'0.44569'         }    {'0.6994'}    {'0.5816'}    {'0.2222'}    {'1'                }    {'0.8738'           }    {'0.8454'           }
    {'Philips_2DEPI_L…'}    {'empty'            }    {'14.1723'          }    {'2.9896'           }    {'0.89531'         }    {'0.7099'}    {'0.5136'}    {'0.1939'}    {'1'                }    {'0.7593'           }    {'0.8910'           }
    {'Philips_2DEPI_S…'}    {'45.9448'          }    {'41.156'           }    {'9.4148'           }    {'0.36505'         }    {'0.8167'}    {'0.4596'}    {'0.1516'}    {'1'                }    {'0.8544'           }    {'0.8028'           }
    {'Philips_3DGRASE…'}    {'35.1573'          }    {'26.2532'          }    {'7.5838'           }    {'0.66854'         }    {'0.5709'}    {'0.5278'}    {'0.4219'}    {'1'                }    {'0.8572'           }    {'0.7547'           }
    {'Philips_3DGRASE…'}    {'32.146'           }    {'24.3005'          }    {'8.5921'           }    {'0.43058'         }    {'0.7239'}    {'0.5520'}    {'0.3564'}    {'1'                }    {'0.8413'           }    {'0.8952'           }
    {'Siemens3DGRASE_…'}    {'37.6936'          }    {'37.2867'          }    {'21.7901'          }    {'0.78873'         }    {'0.5191'}    {'0.4658'}    {'0.4054'}    {'1'                }    {'0.6915'           }    {'0.4987'           }
    {'Siemens_2DEPI_n…'}    {'67.5267'          }    {'50.6622'          }    {'13.4265'          }    {'0.31694'         }    {'0.6887'}    {'0.6031'}    {'0.3101'}    {'1'                }    {'0.8758'           }    {'0.9589'           }`

` 1.8.0 

    {'Data'             }    {'mean_qCBF_TotalGM'}    {'median_qCBF_Tot…'}    {'median_qCBF_Dee…'}    {'CoV_qCBF_TotalGM'}    {'GMvol' }    {'WMvol' }      {'CSFvol'}    {'PipelineCompleted'}    {'TC_ASL_Registra…'}    {'TC_M0_Registrat…'}
    {'GE_3Dspiral_T1p…'}    {'65.2658'          }    {'52.4785'          }    {'31.8834'          }    {'0.29875'         }    {[0.6399]}    {[0.5052]}    {[0.4329]}    {[                1]}    {[           0.9005]}    {[           0.9319]}
    {'Philips_2DEPI_F…'}    {'39.0707'          }    {'27.7337'          }    {'6.7209'           }    {'0.71364'         }    {[0.5349]}    {[0.4587]}    {[0.4051]}    {[                1]}    {[           0.7297]}    {[           0.8139]}
    {'Philips_2DEPI_F…'}    {'74.0677'          }    {'56.3657'          }    {'22.9638'          }    {'0.33674'         }    {[0.5442]}    {[0.3921]}    {[0.3564]}    {[                1]}    {[           0.9191]}    {[           0.9479]}
    {'Philips_2DEPI_F…'}    {'48.8506'          }    {'43.3796'          }    {'20.2502'          }    {'0.41961'         }    {[0.6982]}    {[0.5828]}    {[0.2224]}    {[                1]}    {[           0.8134]}    {[           0.8345]}
    {'Philips_2DEPI_L…'}    {'empty'            }    {'14.436'           }    {'3.4311'           }    {'0.89236'         }    {[0.7099]}    {[0.5136]}    {[0.1939]}    {[                1]}    {[           0.7443]}    {[           0.8896]}
    {'Philips_2DEPI_S…'}    {'45.9493'          }    {'41.1539'          }    {'9.4059'           }    {'0.36499'         }    {[0.8167]}    {[0.4596]}    {[0.1516]}    {[                1]}    {[           0.8543]}    {[           0.8027]}
    {'Philips_3DGRASE…'}    {'35.2382'          }    {'26.316'           }    {'7.6074'           }    {'0.66134'         }    {[0.5709]}    {[0.5279]}    {[0.4218]}    {[                1]}    {[           0.8571]}    {[           0.7547]}
    {'Philips_3DGRASE…'}    {'32.1459'          }    {'24.3005'          }    {'8.5921'           }    {'0.43058'         }    {[0.7239]}    {[0.5520]}    {[0.3564]}    {[                1]}    {[           0.8413]}    {[           0.8952]}
    {'Siemens3DGRASE_…'}    {'37.5377'          }    {'37.2731'          }    {'21.767'           }    {'0.78865'         }    {[0.5191]}    {[0.4658]}    {[0.4054]}    {[                1]}    {[           0.6914]}    {[           0.4987]}
    {'Siemens_2DEPI_n…'}    {'67.5267'          }    {'50.6621'          }    {'13.4265'          }    {'0.31694'         }    {[0.6887]}    {[0.6031]}    {[0.3101]}    {[                1]}    {[           0.8758]}    {[           0.9589]}`

jan-petr commented:
I went through the logs AND:
1. for `Philips_2DEPI_FLAIR_Segm_PWI4D_Quant_Parameters_LoQ`, 1.8 runs ConventionalM0Processing and 1.7 does run the 'new M0 processing'. This is incorrect as M0ConventionalProcessing is set to 1 in the ASL4D.json - but this was probably ignored previously. Also - the warning `Warning: multi-TE processing still work in progress...` should be gone...

Solutions:
1. `Philips_2DEPI_FLAIR_Segm_PWI4D_Quant_Parameters_LoQ` - we test-run in without ConventionalM0Processing option set to 1. If this is similar to 1.7.0, then this was the difference. We then keep the old data and create a new reference.
2. Make sure that `Warning: multi-TE processing still work in progress...` is gone for `Philips_2DEPI_FLAIR_Segm_PWI4D_Quant_Parameters_LoQ` - or have I just processed with older code?! I will make breakpoint to investigate.

MichaelStritt commented:
@jan-petr

2. I reran the ASL module with 1.7.0 and 1.8.0 for `Philips_2DEPI_FLAIR_Segm_PWI4D_Quant_Parameters_LoQ`. In 1.8.0 the `M0_conventionalProcessing` option seems to work and in 1.7.0 it does not. What I do not get is why the option is embedded in the `ASL4D.json` file. That's a processing setting and should be in `dataPar.json`. For me it worked in 1.7.0 correctly if I move the `M0_conventionalProcessing` option to `dataPar.json`.

MichaelStritt commented:
@jan-petr 

I added a small fix for the bug shown above. Didn't want to overstep any boundaries again, which is why I created this separate PR into release https://github.com/ExploreASL/ExploreASL/pull/819 .

jan-petr commented:
@MichaelStritt I have tried to only modify 1.8.0 version - modify the JSONs to see if we can reproduce 1.7.0 results from above and wasn't really successful:

### Philips_2DEPI_Lesion_NoFLAIR_NoM0_HiQ
Without BASIL, there is no difference between 1.7.0 and 1.8.0. But the strange thing is that BASIL is outputted in a separated CBF_Basil.nii, so it shouldn't have affected any of the results previously. Where does the previous different number come from then if now is all good? I was getting median_CBF_Total_GM_B 14.4 all the time for this dataset - how did you manage to increase it from 14.1 to 14.4 for 1.7.0?

### Philips_2DEPI_FLAIR_Segm_PWI4D_Quant_Parameters_LoQ
`"M0_conventionalProcessing":1` - this indeed needs to be removed as this was previously wrong.
These fields appeared to be used correctly also before so I have kept them:
`"Q":{"BackGrSupprPulses":2, "LabelingType":"CASL", "InitialPostLabelDelay":1.800, "LabelingDuration":1.800, "BloodT1":1.818,"Lambda":0.9, "T2":0.165,"TissueT1":1.300, "nCompartments":2, "SliceReadoutTime":"shortestTR"}`
But this didn't arrive to the same results as in 1.7. 
So I have tried to remove additionally these:
`"bAffineRegistration":2, , "M0":"separate_scan", ` but that didn't help either completely...

### Philips_2DEPI_FLAIR_DisabledQuantification_M0_LoQ
If I understand it correctly. I can only get the latest release branch and it should work without modifying the JSONs? Was there an issue with a vertical vector ApplyQUantification== [0 0 0 0 ] being considered as a single 0 and being supplemented as [0 1 1 1] or something like that?
The trouble is that upon re-running with the bugfix I have still the same high values...

Sorry - can't really approve, until I understand everything and manage to re-run the 1.8.0 to somehow make it fit to 1.7.0 without re-running 1.7.0. We can discuss it tomorrow.

MichaelStritt commented:
This 86591ae isn't a good idea. We should never check a variable in line 138 that is possibly not defined. I know that this condition will not be checked because of bUseBasilQuantification would be true in that case, but still it's not really nice coding if we rely on that...

In addition I was testing **everything** based on the state from yesterday evening... If you want those two edits in the release, we would have to retest **everything** again... Is that really necessary???

MichaelStritt commented:
### Test results from before 084349b, 86591ae & 14913c3 

Everything fine here...

```
====================================== TEST RESULTS ======================================

                        name                           unit           tests        passed
    ____________________________________________    __________    _____________    ______

    'xASL_adm_CatchNumbersFromString'               'function'    [1×3  struct]    true  
    'xASL_adm_CheckFileCount'                       'function'    [1×2  struct]    true  
    'xASL_adm_CompareLists'                         'function'    [1×2  struct]    true  
    'xASL_adm_CorrectName'                          'function'    [1×3  struct]    false 
    'xASL_adm_DeleteFileList'                       'function'    [1×2  struct]    true  
    'xASL_adm_OrderFields'                          'function'    [1×1  struct]    true  
    'xASL_bids_BIDS2Legacy'                         'function'    [1×2  struct]    true  
    'xASL_bids_CheckDatasetDescription'             'function'    [1×2  struct]    true  
    'xASL_bids_Config'                              'function'    [1×1  struct]    true  
    'xASL_bids_CreateDatasetDescriptionTemplate'    'function'    [1×2  struct]    true  
    'xASL_bids_JsonCheck'                           'function'    [1×3  struct]    true  
    'xASL_bids_VendorFieldCheck'                    'function'    [1×1  struct]    true  
    'xASL_io_ExportVTK'                             'function'    [1×5  struct]    true  
    'xASL_io_Nifti2Im'                              'function'    [1×1  struct]    true  
    'xASL_stat_PSNR'                                'function'    [1×2  struct]    true  
    'xASL_stat_QuantileNan'                         'function'    [1×1  struct]    true  
    'xASL_stat_StdNan'                              'function'    [1×3  struct]    true  
    'xASL_str2num'                                  'function'    [1×3  struct]    true  
    'xASL_test_GetLogContent'                       'function'    [1×2  struct]    true  
    'xASL_tsvRead'                                  'function'    [1×2  struct]    true  
    'xASL_tsvWrite'                                 'function'    [1×2  struct]    true  
    'xASL_ExploreASL_Master'                        'master'      [1×11 struct]    true  
    'xASL_module_ASL'                               'module'      [1×1  struct]    true  
    'xASL_module_Population'                        'module'      [1×1  struct]    true  
    'xASL_module_Structural'                        'module'      [1×1  struct]    true  

==========================================================================================
```

MichaelStritt commented:
@HenkMutsaerts: let's discuss this here, because whatsapp isn't really that good for this kind of stuff

084349b and 14913c3 are okay, but 86591ae definitely isn't. I'm just saying that we should change...

```
    if bUseBasilQuantification
        % Here we perform FSL BASIL
        [PWI, resultFSL] = xASL_quant_Basil(PWI, x);
    end
```

...back to...

```
    if bUseBasilQuantification
        % Here we perform FSL BASIL
        [PWI, resultFSL] = xASL_quant_Basil(PWI, x);
    else
        resultFSL = NaN;
    end
```

I can't use the windows machine to test everything again today and I'm not in the office tomorrow, but somebody can test everything on linux and then it should basically be fine, because we had multiple testing rounds already. Agree @jan-petr & @HenkMutsaerts ?

MichaelStritt commented:
> @MichaelStritt can you please undo this, I already commented before that `xASL_adm_CreateDir` automatically checks if it has to create a folder yes or no right?

But I don't want to have an arbitrary warning (folder exists already) at this point. I want to warn the user that this specific BIDS subject exists already. There's a difference there.

AND

I know you want to do this in more detail and discuss it and so on, but it is helpful to have it in 1.8.0 and it was tested a lot. That's just not moving in the right direction and I don't have time for this kind of stuff.

jan-petr commented:
One more try during the day ;)

### Philips_2DEPI_FLAIR_DisabledQuantification_M0_LoQ
Yes, the problem was probably that the quantification wasn't skipped. If we skip it, we obtain similar numbers. Managed to reproduce the old 1.7.0 with 1.8.0. We are thus fine with the 2-fold increase in CBF.

The other two are probably also fine - one was caused by the M0Processing and the other by the other parameters. But I guess we can fix the values now. But we should then save the 1.8.0 values as reference for both Windows and Linux.

MichaelStritt commented:
> 
> 
> One more try during the day ;)
> ### Philips_2DEPI_FLAIR_DisabledQuantification_M0_LoQ
> 
> Yes, the problem was probably that the quantification wasn't skipped. If we skip it, we obtain similar numbers. Managed to reproduce the old 1.7.0 with 1.8.0. We are thus fine with the 2-fold increase in CBF.
> 
> The other two are probably also fine - one was caused by the M0Processing and the other by the other parameters. But I guess we can fix the values now. But we should then save the 1.8.0 values as reference for both Windows and Linux.

Sounds good. How do we handle the BASIL dataset?

MichaelStritt commented:
### Windows test datasets

![image](https://user-images.githubusercontent.com/46593074/132804395-cb3604a4-ba9e-43e0-99b7-f3e3a2272dbe.png)

HenkMutsaerts commented:
* You know that you should be able to copy paste the output you're showing in that Table to this Github issue? For me this always gives a very nice layout on Github automatically
* Why do we have both #805 and #816? :)
* I will test now, having FSL. I guess we should issue a proper warning `if bUseBasilQuantification && ` FSL is not found, and then set `bUseBasilQuantification` to `false`. Then the pipeline doesn't crash if you request Basil but don't have it. We would have to test the same for TopUp (first step of the ASL module) where you also use FSL.
* We discussed already with the FSL developers, best would be to include a light-weight version of FSL in ExploreASL
*

jan-petr commented:
@MichaelStritt Yes, it definitively needs to be fixed. For once, it always does the normal quantification as well, even if BASIL is not installed, so it should never crash and always deliver results.

HenkMutsaerts commented:
I get this error, any ideas?

```
ExploreASL will run the import workflow and will load the dataset...
==============================================================================================
 ________                      __                                 ______    ______   __        
/        |                    /  |                               /      \  /      \ /  |      
########/  __    __   ______  ## |  ______    ______    ______  /######  |/######  |## |      
## |__    /  \  /  | /      \ ## | /      \  /      \  /      \ ## |__## |## \__##/ ## |      
##    |   ##  \/##/ /######  |## |/######  |/######  |/######  |##    ## |##      \ ## |      
#####/     ##  ##<  ## |  ## |## |## |  ## |## |  ##/ ##    ## |######## | ######  |## |      
## |_____  /####  \ ## |__## |## |## \__## |## |      ########/ ## |  ## |/  \__## |## |_____ 
##       |/##/ ##  |##    ##/ ## |##    ##/ ## |      ##       |## |  ## |##    ##/ ##       |
########/ ##/   ##/ #######/  ##/  ######/  ##/        #######/ ##/   ##/  ######/  ########/ 
                    ## |                                                                      
                    ## |                                                                      
                    ##/  

==================================== ExploreASL Settings =====================================
Dataset Root        ...ers/henk/ExploreASL/TmpConversion/Philips_PCASL_3DGRASE_5.4.1.0_TopUp_1
Import Modules      DCM2NII 
Process Modules     
bPause              False
iWorker             1
nWorkers            1
==============================================================================================
ExploreASL v1.8.0 initialized ... 
================================== DICOM to NIFTI CONVERSION =================================

Matching files (#=2):
Sub1/ASL
Sub1/M0

Running DCM2NIIX...
==============================================================================================
Importing subject=Sub1_1:   
>>> Subject=Sub1, visit=1, session=1, scan=ASL4D
executing: [/Users/henk/ExploreASL/ExploreASL/External/MRIcron/20190902/dcm2nii-osx -o /Users/henk/ExploreASL/TmpConversion/Philips_PCASL_3DGRASE_5.4.1.0_TopUp_1/temp/Sub1/ASL_1/dcm2nii_temp_ASL4D /Users/henk/ExploreASL/TmpConversion/Philips_PCASL_3DGRASE_5.4.1.0_TopUp_1/sourcedata/Sub1/ASL/00001.dcm]
Chris Rorden's dcm2niiX version v1.0.20190902  (JP2:OpenJPEG) (JP-LS:CharLS) Clang8.1.0 (64-bit MacOS)
Found 10 DICOM file(s)
Warning: Empty protocol name(s) (0018,1030)
Warning: Unable to append protocol name (0018,1030) to filename (it is empty).
Using RWVSlope:RWVIntercept = 0.1:0
 Philips Scaling Values RS:RI:SS = 0.1:0:10 (see PMC3998685)
Convert 10 DICOM as /Users/henk/ExploreASL/TmpConversion/Philips_PCASL_3DGRASE_5.4.1.0_TopUp_1/temp/Sub1/ASL_1/dcm2nii_temp_ASL4D/ASL__20181204173011_801 (80x80x10x1)
Conversion required 0.144819 seconds (0.023110 for core code).
==============================================================================================
Chris Rorden's dcm2niiX version v1.0.20190902  (JP2:OpenJPEG) (JP-LS:CharLS) Clang8.1.0 (64-bit MacOS)
Found 10 DICOM file(s)
Warning: Empty protocol name(s) (0018,1030)
Warning: Unable to append protocol name (0018,1030) to filename (it is empty).
Using RWVSlope:RWVIntercept = 0.1:0
 Philips Scaling Values RS:RI:SS = 0.1:0:10 (see PMC3998685)
Convert 10 DICOM as /Users/henk/ExploreASL/TmpConversion/Philips_PCASL_3DGRASE_5.4.1.0_TopUp_1/temp/Sub1/ASL_1/dcm2nii_temp_ASL4D/ASL__20181204173011_801 (80x80x10x1)
Conversion required 0.144819 seconds (0.023110 for core code).
==============================================================================================
status: 0
Recreating parameter file: /Users/henk/ExploreASL/TmpConversion/Philips_PCASL_3DGRASE_5.4.1.0_TopUp_1/temp/Sub1/ASL_1/ASL4D.js100%Field ImageType contains empty fields, skipping
>>> Subject=Sub1, visit=1, session=1, scan=M0
executing: [/Users/henk/ExploreASL/ExploreASL/External/MRIcron/20190902/dcm2nii-osx -o /Users/henk/ExploreASL/TmpConversion/Philips_PCASL_3DGRASE_5.4.1.0_TopUp_1/temp/Sub1/ASL_1/dcm2nii_temp_M0 /Users/henk/ExploreASL/TmpConversion/Philips_PCASL_3DGRASE_5.4.1.0_TopUp_1/sourcedata/Sub1/M0/AP0]
Chris Rorden's dcm2niiX version v1.0.20190902  (JP2:OpenJPEG) (JP-LS:CharLS) Clang8.1.0 (64-bit MacOS)
Found 38 DICOM file(s)
Warning: Empty protocol name(s) (0018,1030)
Warning: Unable to append protocol name (0018,1030) to filename (it is empty).
Using RWVSlope:RWVIntercept = 1.31746:0
 Philips Scaling Values RS:RI:SS = 1.31746:0:0.00291533 (see PMC3998685)
Convert 19 DICOM as /Users/henk/ExploreASL/TmpConversion/Philips_PCASL_3DGRASE_5.4.1.0_TopUp_1/temp/Sub1/ASL_1/dcm2nii_temp_M0/M0__20181204173011_601 (80x80x19x1)
Warning: Unable to append protocol name (0018,1030) to filename (it is empty).
Using RWVSlope:RWVIntercept = 1.3011:0
 Philips Scaling Values RS:RI:SS = 1.3011:0:0.00259727 (see PMC3998685)
Convert 19 DICOM as /Users/henk/ExploreASL/TmpConversion/Philips_PCASL_3DGRASE_5.4.1.0_TopUp_1/temp/Sub1/ASL_1/dcm2nii_temp_M0/M0__20181204173011_701 (80x80x19x1)
Conversion required 0.394456 seconds (0.065465 for core code).
==============================================================================================
Chris Rorden's dcm2niiX version v1.0.20190902  (JP2:OpenJPEG) (JP-LS:CharLS) Clang8.1.0 (64-bit MacOS)
Found 38 DICOM file(s)
Warning: Empty protocol name(s) (0018,1030)
Warning: Unable to append protocol name (0018,1030) to filename (it is empty).
Using RWVSlope:RWVIntercept = 1.31746:0
 Philips Scaling Values RS:RI:SS = 1.31746:0:0.00291533 (see PMC3998685)
Convert 19 DICOM as /Users/henk/ExploreASL/TmpConversion/Philips_PCASL_3DGRASE_5.4.1.0_TopUp_1/temp/Sub1/ASL_1/dcm2nii_temp_M0/M0__20181204173011_601 (80x80x19x1)
Warning: Unable to append protocol name (0018,1030) to filename (it is empty).
Using RWVSlope:RWVIntercept = 1.3011:0
 Philips Scaling Values RS:RI:SS = 1.3011:0:0.00259727 (see PMC3998685)
Convert 19 DICOM as /Users/henk/ExploreASL/TmpConversion/Philips_PCASL_3DGRASE_5.4.1.0_TopUp_1/temp/Sub1/ASL_1/dcm2nii_temp_M0/M0__20181204173011_701 (80x80x19x1)
Conversion required 0.394456 seconds (0.065465 for core code).
==============================================================================================
status: 0
Recreating parameter file: /Users/henk/ExploreASL/TmpConversion/Philips_PCASL_3DGRASE_5.4.1.0_TopUp_1/temp/Sub1/ASL_1/M0_601_00601.json
Recreating parameter file: /Users/henk/ExploreASL/TmpConversion/Philips_PCASL_3DGRASE_5.4.1.0_TopUp_1/temp/Sub1/ASL_1/M0_701_00701.js100%Field ImageType contains empty fields, skipping
Field ImageType contains empty fields, skipping



Error using xASL_Move (line 48)
Source doesnt exist:
/Users/henk/ExploreASL/TmpConversion/Philips_PCASL_3DGRASE_5.4.1.0_TopUp_1/temp/Sub1/ASL_1/M0_601_00001.nii

Error in xASL_test_Flavors_DCM2BIDS (line 67)
			xASL_Move(fullfile(DirASL, 'M0_601_00001.nii'), fullfile(DirASL,
                        'M0.nii'), 1);

Error in xASL_test_Flavors (line 127)
	xASL_test_Flavors_DCM2BIDS(conversionPath, x);
 
```

HenkMutsaerts commented:
Testing on macOS, with Jan, on Friday-eve trying to find out what is going on...

Someone decided to use Matlab's built in `dir` without explicitly using the `-s` command for recursive, or even using our own `xASL_adm_GetFileList`; and if Matlab is not higher than 2016 (my version! ;) than it gives the following cryptical `fprintf` to the screen:

`This method is not able to find empty directories right now...`

MichaelStritt commented:
> * Why do we have both [Release 1.8.0 #805](https://github.com/ExploreASL/ExploreASL/issues/805) and [Release 1.8.0 #816](https://github.com/ExploreASL/ExploreASL/pull/816)? :)

The first one is an issue, the second one is this PR right here that you're commenting in.

> * We discussed already with the FSL developers, best would be to include a light-weight version of FSL in ExploreASL

I thought the conclusion was that this wouldn't work for a commercial version.

MichaelStritt commented:
@jan-petr & @HenkMutsaerts:

- For the two BASIL datasets in the 10 test datasets I had to change the BASIL fallback/default. Returning an empty image has to lead to a pipeline crash. Instead we could return the original/backup PWI. I implemented this [here](https://github.com/ExploreASL/ExploreASL/commit/0df76a1e7bf734ebbac066626f7699a12b51052b).
- `Philips_PASL_2DEPI_5.4.1.0_ADNI3_006_S_6681_1` crashed just because the corresponding fields were missing in Q. [Here](
https://github.com/ExploreASL/ExploreASL/commit/bcf7aece94ab7835c6346f868c64858dc628c96c) I just check for their existence. Like this it runs the full pipeline without crashing already. Should they exist though? Another big issue here is that we get this warning `Field SliceReadoutTime in xASL structure has a value of 41975, which is outside of the recommended range` even though we explicitly define the SliceTiming of 41.975. Easy to see that we scale ms to seconds here. So we just need to use `0.041975` in the studyPar JSON instead of `"SliceTiming": 41.975`?

MichaelStritt commented:
> 
> 
> Testing on macOS, with Jan, on Friday-eve trying to find out what is going on...
> 
> Someone decided to use Matlab's built in `dir` without explicitly using the `-s` command for recursive, or even using our own `xASL_adm_GetFileList`; and if Matlab is not higher than 2016 (my version! ;) than it gives the following cryptical `fprintf` to the screen:
> 
> `This method is not able to find empty directories right now...`

It's a lot easier to fix these things if you let us know where is `dir` was used.

MichaelStritt commented:
> Testing on macOS, with Jan, on Friday-eve trying to find out what is going on...
> 
> Someone decided to use Matlab's built in `dir` without explicitly using the `-s` command for recursive, or even using our own `xASL_adm_GetFileList`; and if Matlab is not higher than 2016 (my version! ;) than it gives the following cryptical `fprintf` to the screen:
> 
> `This method is not able to find empty directories right now...`

This has nothing to do with `dir` not working correctly. I originally wrote the code on Matlab 2019a and Jan also always used 2019b or newer. Once Mathijs wanted to test with an older version and mentioned that it didn't work for him. Therefore we added this code section below for backwards compatibility...

```
    % Get files and folders of datasets A and B
    [~, dateVersion] = version;
    if xASL_str2num(dateVersion(end-4:end))>2016 % dir does not work recursively in older versions
        filesA = dir(fullfile(pathDatasetA, '**','*.*'));
        filesB = dir(fullfile(pathDatasetB, '**','*.*'));
    else
        [filesA, filesB] = xASL_bids_CompareStructures_GetFileListsUnix(pathDatasetA,pathDatasetB);
    end
```

... the fallback function for older versions `is not able to find empty directories right now` though, which is why we added this print statement. For us this wasn't super relevant at the time, since only me and Jan did the flavor testing so far. Sometimes you have to find a good trade of there. If you know a nice way to make this work, feel free to improve the code 👍 

I don't think the bids comparison function is the problem here though, I think our code tries to move an M0 that's not there. I guess something with the TopUp dataset doesn't work correctly in the import? Those are work in progress as far as I know, but please correct me if I'm wrong here @jan-petr & @HenkMutsaerts 👍

MichaelStritt commented:
### Windows 10 test datasets

![image](https://user-images.githubusercontent.com/46593074/133037115-74385f80-a26d-46ab-a4d3-dba4066fad5c.png)

![image](https://user-images.githubusercontent.com/46593074/133037242-1cc0d652-8b82-4a5b-b800-5fc0b84deeb3.png)

MichaelStritt commented:
## Flavor testing

### Philips_PCASL_3DGRASE_5.1.7.2_1

```
Warning: ...\TmpConversion\Philips_PCASL_3DGRASE_5.1.7.2_1\derivatives\ExploreASL\sub-Patient1_1
didnt contain a T1w structural image, skipping... 
> In xASL_module_Structural (line 65)

...

ERROR: ASL module terminated for subject 1: sub-Patient1_1
ans =
    'Error using xASL_module_ASL (line 220)
     Invalid ASL image
...
```

### Philips_PCASL_3DGRASE_5.4.0.2_1

```
Warning: Corrupt ...\TmpConversion\Philips_PCASL_3DGRASE_5.4.0.2_1\derivatives\ExploreASL\sub-Sub001_1\T1.nii 
> In xASL_io_Nifti2Im (line 81)

...

------------------------------------------------------------------------
CAT Preprocessing error for T1:
------------------------------------------------------------------------
insufficient image overlap
------------------------------------------------------------------------
  474 - error_message
  177 - spm_affreg
...
```

jan-petr commented:
` Linux Matlab2019b

    {'Data'             }    {'mean_qCBF_TotalGM'}    {'median_qCBF_Tot…'}    {'median_qCBF_Dee…'}    {'CoV_qCBF_TotalGM'}    {'GMvol' }    {'WMvol' }      {'CSFvol'}    {'PipelineCompleted'}    {'TC_ASL_Registra…'}    {'TC_M0_Registrat…'}
    {'GE_3Dspiral_T1p…'}    {'65.2658'          }    {'52.4785'          }    {'31.8834'          }    {'0.29875'         }    {[0.6399]}    {[0.5052]}      {[0.4329]}    {[                1]}    {[           0.9005]}    {[           0.9319]}
    {'Philips_2DEPI_F…'}    {'39.0707'          }    {'27.7337'          }    {'6.7209'           }    {'0.71364'         }    {[0.5349]}    {[0.4587]}     {[0.4051]}    {[                1]}    {[           0.7297]}    {[           0.8139]}
    {'Philips_2DEPI_F…'}    {'74.0677'          }    {'56.3657'          }    {'22.9638'          }    {'0.33674'         }    {[0.5442]}    {[0.3921]}     {[0.3564]}    {[                1]}    {[           0.9191]}    {[           0.9479]}
    {'Philips_2DEPI_F…'}    {'48.8506'          }    {'43.3796'          }    {'20.2502'          }    {'0.41961'         }    {[0.6982]}    {[0.5828]}     {[0.2224]}    {[                1]}    {[           0.8134]}    {[           0.8345]}
    {'Philips_2DEPI_L…'}    {'empty'            }    {'14.436'           }    {'3.4311'           }    {'0.89236'         }    {[0.7099]}    {[0.5136]}     {[0.1939]}    {[                1]}    {[           0.7443]}    {[           0.8896]}
    {'Philips_2DEPI_S…'}    {'45.9493'          }    {'41.1539'          }    {'9.4059'           }    {'0.36499'         }    {[0.8167]}    {[0.4596]}     {[0.1516]}    {[                1]}    {[           0.8543]}    {[           0.8027]}
    {'Philips_3DGRASE…'}    {'35.2382'          }    {'26.316'           }    {'7.6074'           }    {'0.66134'         }    {[0.5709]}    {[0.5279]}     {[0.4218]}    {[                1]}    {[           0.8571]}    {[           0.7547]}
    {'Philips_3DGRASE…'}    {'32.1459'          }    {'24.3005'          }    {'8.5921'           }    {'0.43058'         }    {[0.7239]}    {[0.5520]}     {[0.3564]}    {[                1]}    {[           0.8413]}    {[           0.8952]}
    {'Siemens3DGRASE_…'}    {'37.5377'          }    {'37.2731'          }    {'21.767'           }    {'0.78865'         }    {[0.5191]}    {[0.4658]}     {[0.4054]}    {[                1]}    {[           0.6914]}    {[           0.4987]}
    {'Siemens_2DEPI_n…'}    {'67.5267'          }    {'50.6621'          }    {'13.4265'          }    {'0.31694'         }    {[0.6887]}    {[0.6031]}    {[0.3101]}    {[                1]}    {[           0.8758]}    {[           0.9589]} `

MichaelStritt commented:
## BIDS conversion

Here everything seems to work fine...

```
>> xASL_test_Flavors(pathExploreASL, pathTest, [0 0 1 0 0 0 0], x);
Dataset: GE_PCASL_2DEPI_23.0LX_1
Dataset: GE_PCASL_3Dspiral_14.0LX_1
Dataset: GE_PCASL_3Dspiral_15.0LX_WIP_1
Dataset: GE_PCASL_3Dspiral_22.0LX_1
Dataset: GE_PCASL_3Dspiral_23.1LX_1
Dataset: GE_PCASL_3Dspiral_24.0LX_WIP_1
Dataset: GE_PCASL_3Dspiral_25.0LX_ADNI3_027_S_5079_1
Dataset: Philips_PASL_2DEPI_5.4.1.0_ADNI3_006_S_6681_1
Dataset: Philips_PCASL_2DEPI_3.2.1.1_1
Dataset: Philips_PCASL_2DEPI_3.2.1.1_2
Dataset: Philips_PCASL_2DEPI_3.2.1.1_3
Dataset: Philips_PCASL_2DEPI_3.2.1.1_dummyLL_1
Dataset: Philips_PCASL_2DEPI_3.2.1.1_dummyMultiPLD_1
Dataset: Philips_PCASL_2DEPI_3.2.1.1_dummyQUASAR_1
Dataset: Philips_PCASL_2DEPI_3.2.2.0_glioma_1
Dataset: Philips_PCASL_2DEPI_3.2.2.1_1
Dataset: Philips_PCASL_2DEPI_3.2.3.0_BSupN_1
Dataset: Philips_PCASL_2DEPI_3.2.3.1_1
Dataset: Philips_PCASL_2DEPI_3.2.3.2_1
Dataset: Philips_PCASL_2DEPI_4.1.2.1_1
Dataset: Philips_PCASL_2DEPI_5.1.7.2_1
Dataset: Philips_PCASL_2DEPI_5.1.7.2_2
Dataset: Philips_PCASL_2DEPI_5.1.8.2_1
Dataset: Philips_PCASL_2DEPI_5.1.9.1_GMB_1
Dataset: Philips_PCASL_2DEPI_5.3.1.1_1
Dataset: Philips_PCASL_2DEPI_PARREC_1
Dataset: Philips_PCASL_3DGRASE_5.1.7.2_1
Dataset: Philips_PCASL_3DGRASE_5.4.0.2_1
Warning: Corrupt
M:\SoftwareDevelopment\MATLAB\m.stritt\Server_xASL\FlavorTests\TmpConversion\Philips_PCASL_3DGRASE_5.4.0.2_1\rawdata\sub-Sub001\anat\sub-Sub001_T1w_tmp_copy.nii.gz 
> In xASL_io_Nifti2Im (line 81)
  In xASL_im_CompareNiftis (line 90)
  In xASL_bids_CompareStructures>compareNIFTI (line 590)
  In xASL_bids_CompareStructures>checkFileContents (line 332)
  In xASL_bids_CompareStructures (line 201)
  In xASL_test_Flavors (line 137) 
Message: File is smaller than the dimensions say it should be.
Dataset: Philips_PCASL_3DGRASE_5.4.1.0_TopUp_1
Dataset: Philips_PCASL_3DGRASE_5.7.1.1_classic
Dataset: Siemens_PASL_2DEPI_VB15A_ADNI2_941_S_4764_1
Dataset: Siemens_PASL_2DEPI_VB17A_ADNI2_941_S_5193_1
Dataset: Siemens_PASL_2DEPI_VB17A_BSupN_1
Dataset: Siemens_PASL_2DEPI_VB17A_BSupN_2
Dataset: Siemens_PASL_2DEPI_VB19A_BSupN_1
Dataset: Siemens_PASL_3DGRASE_E11_1
Dataset: Siemens_PASL_3DGRASE_VB17A_1
Dataset: Siemens_PASL_3DGRASE_VD13D_1
Dataset: Siemens_PASL_3DGRASE_VE11C_1
Dataset: Siemens_PASL_3DGRASE_VE11C_2
Dataset: Siemens_PCASL_2DEPI_NA_1
Dataset: Siemens_PCASL_2DEPI_VB17A_BSupN_1
Dataset: Siemens_PCASL_2DEPI_VB17A_BSupN_2
Dataset: Siemens_PCASL_3DGRASE_VB17A_TopUp_1
Dataset: Siemens_PCASL_3DGRASE_VB17A_TopUp_2
Dataset: Siemens_PCASL_3DGRASE_VB17A_multiPLD_1
Dataset: Siemens_PCASL_3DGRASE_VD13A_Hadamard4-FME_1
Dataset: Siemens_PCASL_3DGRASE_VD13A_Hadamard8-FME_1
Dataset: Siemens_PCASL_3DGRASE_VD13D_1
Dataset: Siemens_PCASL_3DGRASE_VD13D_2
Dataset: Siemens_PCASL_3DGRASE_VE11C_1
Dataset: Siemens_PCASL_3DGRASE_VE11C_2
```
